dev/io.openliberty.org.jboss.resteasy.common/src/org/jboss/resteasy/plugins/server/servlet/ServletUtil.java

```diff
@@ -0,0 +1,179 @@
+import org.jboss.resteasy.specimpl.ResteasyUriInfo.InitData;
+import org.jboss.resteasy.util.HttpHeaderNames;
+import org.jboss.resteasy.util.MediaTypeHelper;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.ws.rs.core.Cookie;
+import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.MultivaluedMap;
+
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * @author <a href="mailto:bill@burkecentral.com">Bill Burke</a>
+ * @version $Revision: 1 $
+ */
+public class ServletUtil
+{
+   private static Map<String, InitData> initDataCache = new HashMap<String, InitData>();      // Liberty change
+    
+   public static ResteasyUriInfo extractUriInfo(HttpServletRequest request, String servletPrefix)
+   {
+      String contextPath = request.getContextPath();
+      if (servletPrefix != null && servletPrefix.length() > 0 && !servletPrefix.equals("/"))
+      {
+         if (!contextPath.endsWith("/") && !servletPrefix.startsWith("/"))
+            contextPath += "/";
+         contextPath += servletPrefix;
+      }
+      String queryString = request.getQueryString();
+      String absolute;
+      if (queryString != null && queryString.length() > 0)
+      {
+         absolute = request.getRequestURL().append('?').append(queryString).toString();
+      }
+      else
+      {
+         absolute = request.getRequestURL().toString();
+      }
+      if (!absolute.contains(contextPath))
+      {
+         String encodedContextPath = Arrays.stream(contextPath.substring(1).split("/"))
+                                           .map(s -> {
+                                              try {
+                                                 return URLEncoder.encode(s, "UTF-8");
+                                                } catch (UnsupportedEncodingException ex) {
+                                                 return s;
+                                              }
+                                             })
+                                           .collect(Collectors.joining("/", "/", ""));
+         if (absolute.contains(encodedContextPath))
+         {
+            absolute = absolute.replace(encodedContextPath, contextPath);
+         }
+      }
+      // Liberty change begin      
+      ResteasyUriInfo myResteasyUriInfo;
+      if(InitData.canBeCached(absolute)) {
+          String cacheKey = absolute + contextPath;
+          InitData myInitData = initDataCache.get(cacheKey);
+          if(myInitData == null) {
+              myInitData = new InitData(absolute, contextPath);
+              initDataCache.put(cacheKey, myInitData);
+          }
+          myResteasyUriInfo = new ResteasyUriInfo(absolute, contextPath, myInitData);
```

Comment:
The javadoc for this method states:
```
    * Meant to be called by a layer that know that the absoluteUri will result in the successful
    * invocation of a JAX-RS endpoint.
```
Does this need to be validated, or is it already validated at this point?  This gets back to the memory leak question.  If we put in garbage can we end up caching a lot of things leading to a memory leak.

Reply:
I don't know how to validate "will result in successful invocation of a JAX-RS endpoint" ... but we already call the two-arg ResteasyUriInfo constructor and it always calls this constructor with "null" for InitData so in general it seems safe enough to call this constructor directly.

With our test app, "absolute" looks like:  http://10.16.112.38:9080/pingApp  - scheme, server, port, and URI are all retrieved from the Liberty classes that process the inbound request, so it doesn't look like we should end up with garbage in the cache.